src/machine/aot/mod.rs

```diff
@@ -0,0 +1,673 @@
+        Value::Register(24),
+        Value::Register(25),
+        Value::Register(26),
+        Value::Register(27),
+        Value::Register(28),
+        Value::Register(29),
+        Value::Register(30),
+        Value::Register(31),
+    ]
+}
+
+struct LabelGatheringMachine {
+    registers: [Value; 32],
+    pc: Value,
+    labels_to_test: Vec<u64>,
+
+    // A memory segment which contains code loaded from ELF
+    memory: FlatMemory<u64>,
+    labels: HashSet<u64>,
+    sections: Vec<(u64, u64)>,
+    dummy_sections: HashMap<u64, u64>,
+}
+
+impl LabelGatheringMachine {
+    pub fn load(program: &Bytes) -> Result<Self, Error> {
+        let elf = Elf::parse(&program).unwrap();
+        if elf.section_headers.len() > MAXIMUM_SECTIONS {
+            return Err(Error::LimitReached);
+        }
+        let mut sections: Vec<(u64, u64)> = elf
+            .section_headers
+            .iter()
+            .filter_map(|section_header| {
+                if section_header.sh_flags & u64::from(SHF_EXECINSTR) != 0 {
+                    Some((
+                        section_header.sh_addr,
+                        section_header.sh_addr + section_header.sh_size,
+                    ))
+                } else {
+                    None
+                }
+            })
+            .rev()
+            .collect();
+        // Test there's no empty section
+        if sections.iter().any(|(s, e)| s >= e) {
+            return Err(Error::OutOfBound);
+        }
+        // Test no section overlaps with one another. We first sort section
+        // list by start, then we test if each end is equal or less than
+        // the next start.
+        sections.sort_by_key(|section| section.0);
+        if sections.windows(2).any(|w| w[0].1 > w[1].0) {
+            return Err(Error::OutOfBound);
+        }
+        // DefaultCoreMachine is only used here for loading ELF binaries
+        // into memory.
+        let mut inner = DefaultCoreMachine::default();
+        inner.load_elf(&program, false)?;
+
+        Ok(Self {
+            registers: init_registers(),
+            pc: Value::from_u64(0),
+            labels: HashSet::default(),
+            labels_to_test: Vec::new(),
+            memory: inner.take_memory(),
+            sections,
+            dummy_sections: HashMap::default(),
+        })
+    }
+
+    fn read_pc(&self) -> Result<u64, Error> {
+        match &self.pc {
+            Value::Imm(pc) => Ok(*pc),
+            _ => Err(Error::Unexpected),
+        }
+    }
+
+    pub fn gather(&mut self) -> Result<(), Error> {
+        let decoder = build_imac_decoder::<u64>();
+        for i in 0..self.sections.len() {
+            let (section_start, section_end) = self.sections[i];
+            self.pc = Value::from_u64(section_start);
+            let mut start_of_basic_block = true;
+            while self.read_pc()? < section_end {
+                let pc = self.read_pc()?;
+                if start_of_basic_block {
```

Comment:
I think should be moved `Ok(instruction)` branch, otherwise, it will add dummy section to labels.

Reply:
I think the result is the same since labels for dummy section will be cleaned, but I agree moving this will make the logic cleaner.